### PR TITLE
[Don't merge, review only] 1.58bit implementation

### LINKED ISSUE #56: Implement 1.58 bit BitDistiller
## 1. Description

At the time BitDistiller was released, the 1.58 bit LLM had not been implemented. 1.58 bit provides a quantisation level between 2 bit and 1 bit. Implementing this is non-trivial as the quantisation/clipping functions of BitDistiller do not natively support 1.58 bit.

Ternary weight quantisation from reference [2]
```
def weight_quant(w):
    scale = 1.0 / w.abs().mean().clamp_(min=1e-5)
    u = (w * scale).round().clamp_(-1, 1) / scale
    return u
```

TODO list:
- [x] Adapt ternary quantisation for `qlinear` BitDistiller Layer (see hugging face link, can ignore the activation quant for now)
- [x] Understand and check if clipping/other architectural changes (see 1.58 bitnet paper and comment below) are needed.
- [x] Get training to run successfully on dry run
- [x] Check weight distribution looks somewhat close to 2 bit, but i didnt' check layer wise. 
- [x] Check if inference modifications needed
- [x] Add 1.58 implementation closer to BitDistiller, see 1.58-range-scale branch, calculates scale using range instead of mean like bitnet 1.58.
- [x] Compare implementations and eval on 2 epochs 
- [x] Full train run on the range scale, same hyperparams as BitDistiller
- [x] Eval and upload model

## 2. Hypothesis

Given the better performance of 1.58 bit vs 1 bit LLMs in pretraining, we should expect to also a significant boost in performance. If the architectural changes are small, we should expect 1.58 bit to perform between 1 and 2 bits in terms of benchmarks.

## 3. Results

[Closer to bitnet 1.58 implementation](https://huggingface.co/BrownianNotion/TinyLlama_v1.1_1.58bit_ternary)
[Closer to BitDistiller implementation](https://huggingface.co/BrownianNotion/TinyLlama_v1.1_1.58bit_ternary_range_scale) 

[Second implementation trained on 4 epochs](https://huggingface.co/BrownianNotion/TinyLlama_v1.1_1.58bit_ternary_range_scale_4epochs)

**Wandb run name:** TinyLlama_v1.1_1.58_ternary_2025-04-02 21:34:19.441882  

For model trained on 4 epochs, even thought the (eval) loss and PPL decreased, the metrics barely changed. No benchmark appears statistically different from 1 bit, which is pretty much as bad as the 1 bit. Not obvious what could be going wrong, haven't deep dived anything yet.

## References
1. https://arxiv.org/pdf/2402.17764
2. https://huggingface.co/blog/1_58_llm_extreme_quantization 

**Reviewer:** @fredericowieser

### PR DESCRIPTION
**Pull request by BrownianNotion**
## Description  
  
Results here https://huggingface.co/BrownianNotion/TinyLlama_v1.1_1.58bit_ternary_range_scale_4epochs

No statistical difference with 1 bit implementation. Based on loss curves the model seems to be learning something but no benchmarks are statistically different from random.

Closes #56

### CODE REVIEW

File: quantization/quantizer.py [hunk 1]
```diff
@@ -258,6 +261,35 @@ def forward(self, x):
...
+        assert x.dim() == 2
+
+        max_val = x.amax(dim=1, keepdim=True)  # (num_groups, 1)
+        min_val = x.amin(dim=1, keepdim=True)  # (num_groups, 1)
+
+        scales = (max_val - min_val).clamp(min=1e-5)
+
+        assert torch.isnan(scales).sum() == 0
+        assert torch.isnan(x).sum() == 0
+
```
Comment (fredericowieser) [inline quantization/quantizer.py:284]: This class looks fine but seems to have a different layout of steps to the methods for 1bit and 2bit. As long as we are sure that this method is correct, then I am happy with it :)

**Review by fredericowieser — COMMENTED** [review]
Looks good overall, and seems to be pretty clean code. No real notes to give :)

File: quantization/quantizer.py (hunk 1 above)
Comment (BrownianNotion) [reply quantization/quantizer.py:284]: can you be more specific about the differences in the layout? Would be good to check in case I missed something. I tried to follow the original class as closely as possible where it made sense.
